libbpf-tools/klockstat.bpf.c

```diff
@@ -0,0 +1,276 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright (c) 2021 Google LLC.
+ *
+ * Based on klockstat from BCC by Jiri Olsa and others
+ * 2021-10-26   Barret Rhoden   Created this.
+ */
+#include "vmlinux.h"
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "klockstat.h"
+#include "bits.bpf.h"
+#include "core_fixes.bpf.h"
+
+const volatile pid_t targ_tgid = 0;
+const volatile pid_t targ_pid = 0;
+struct mutex *const volatile targ_lock = NULL;
+
+/*
+ * Make sure all programs are loaded before we start tracing so that we see
+ * lock/lock_acquried/unlock for all locks.  Otherwise, the bookkeeping can be
```

Comment:
Don't think this is right. Even if we have all BPF programs loaded, we can also see partial lock/unlock event sequence.
What we often do is filter out those event, like: https://github.com/iovisor/bcc/blob/4d08181d194bc2961f5d3e599f6db5d870648d15/tools/tcpconnect.py#L164-L166

Reply:
i think the issue is that the lock-acquired or lock-unlocked events are getting paired up with the wrong lock-contended, possibly due to a mismatch with the depth.  the code already aborts if it can't find a lockholder_info for a given depth, but the issues could arise if you use the wrong depth.  

TBH, i haven't traced out the exact line sequence of events.  i see a bunch of locks grabbed in the kernel when the bpf program is inserted, but i don't know exactly when the fentry/fexit programs start to run in all of that code.

FWIW, the original klockstat.py also used an enabled flag, so they might have had the same issue, especially considering this takes the same approach (tracking via task->depth).   though they also used it when extracting stats.

Reply:
here's what i think happens: if you trace mutex_lock, but not unlock, and then grab and release a bunch of locks, then we'll increment task->depth a bunch of times.  once we start tracing unlock, then all future locks will be paired with an unlock.  those initial unpaired struct lockholder_info will just sit around, and the task->depth will always appear to be holding more locks than it does.

eventually, we'll have some mutex_unlock that wasn't paired with a traced call.  in my original code, this was from mutex_trylock.  there are some other places in the kernel where you can grab a mutex without calling mutex_lock or trylock.  when those callsites unlock, we'll trace it and pair it up with one of those initial locks from way back when klockstat set up.  that's how i was able to get a trace where it thought a lock was held for several seconds.

once a task->depth hits 0, we'll ignore any extra 'unbalanced' unlocks.  it's possible that we'll get some incorrect results, such as 
  mutex_lock X
    weird_untraced_mutex_lock Y
    mutex_unlock Y
  mutex_unlock X

the unlock of Y will be paired with the lock of X, and the unlock of X will be ignored.  this is better than having that unlock of X pair with some mutex_lock from a long time ago.